### ISSUE TITLE
memory leaks with monitors

### ISSUE THREAD

**Issue by ashwinr** [cujojs/when#194]
Looks like the monitor code is holding references in memory indefinitely. This is causing us to leak a lot of memory. Please look at the attached screenshot.

![image](https://f.cloud.github.com/assets/362990/1117917/777424f0-1a55-11e3-9355-3463e91cfc2d.png)

**Comment by rkaw92** [comment]
Hi. I'm seeing a very similar problem on when 3.7.7.
This is a minimal test case, which resembles the code that I've been having problems with in production:

```js
var when = require('when');
require('when/monitor/console');

function read() {
	return when.resolve().delay(0).then(function() {
		callRead();
	});
}

function callRead() {
	read();
}

callRead();
```

It simulates a ReadableStream's control flow, where `callRead()` would typically be the `.push()` to the stream's internal buffer (and `.read()` is automatically called again in flowing mode). This resemblance may be unimportant, but the general case is that it leaks memory when using Promises as a trampoline for recursion.

The code's memory usage goes up to arbitrary numbers of megabytes, until the GC kicks in. Commenting out the `require('when/monitor/console')` makes the process float from 25MB to 32MB and back. Tested on Node 4.4.2, though I have been tracking this memory leak since we deployed our app on 0.10 and a few versions of `when` earlier.

I see two possibilities: either the code really is leaking references to Promises, and the console monitor only brings about the demise of the process faster, or it's the monitor who is leaking *something* (I've only seen Promise references in Chrome inspector so far). Will keep watching the "stable" process for a few hours to determine if there is a memory leak without the monitor module, but so far, it seems there is none.
